[JSC] Add JIT support for RegExp ignore case Backreferences in 16-bit patterns

msaboff opened the pull request:
#### 9ff5c9b7dce64de5003500c07075f884f0e14a32
<pre>
[JSC] Add JIT support for RegExp ignore case Backreferences in 16-bit patterns
<a href="https://bugs.webkit.org/show_bug.cgi?id=271617">https://bugs.webkit.org/show_bug.cgi?id=271617</a>
<a href="https://rdar.apple.com/125333123">rdar://125333123</a>

Reviewed by Yusuke Suzuki.

Added code to call out to a new operationAreCanonicallyEquivalent() C++ helper, which wraps the inlined areCanonicallyEquivalent().
Given that the Yarr JIT has fixed assigned registers, added a thunk that saves all caller saves, moves the arguments from the
Yarr fixed registers to what operationAreCanonicallyEquivalent() expects and then calls it.  Upon return from the operation, we
restore registers, put the comparison return in a known register and return back to the Yarr JIT code that called out.

This change supports ARM64 and X86_64 (non-Windows) only.  Work should only be needed for the coding of the thunk generator for
other platforms with sufficient registers for general backrefernce processing in the Yarr JIT.

* Source/JavaScriptCore/yarr/YarrJIT.cpp:
(JSC::Yarr::JSC_DEFINE_JIT_OPERATION):
(JSC::Yarr::jitCompileInlinedTest):
* Source/JavaScriptCore/yarr/YarrJIT.h:
* Source/WTF/wtf/PlatformEnable.h:

Canonical link: <a href="https://commits.webkit.org/276681@main">https://commits.webkit.org/276681@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d9fc5dbde35016d19d9f0398a662985b8bab2d63

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/45317 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/24439 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/47844 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/47981 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/41325 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/28644 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/21834 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/37163 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/45895 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/21492 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/39090 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18264 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/18901 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/40174 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/3364 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/38537 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/41587 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/40493 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/49704 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/44786 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/20300 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/16828 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/44214 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/21607 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/43023 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10079 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/21967 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/51945 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/21295 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/10606 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/yarr/YarrJIT.cpp:
```diff
@@ -4509,8 +4570,59 @@ class YarrGenerator final : public YarrJITInfo {
+
+        LinkBuffer patchBuffer(jit, GLOBAL_THUNK_ID, LinkBuffer::Profile::YarrJIT);
+
+        patchBuffer.link<OperationPtrTag>(functionCall, operationAreCanonicallyEquivalent);
```
We do not need this.

msaboff replied:
Fixed.

Constellation commented on Source/JavaScriptCore/yarr/YarrJIT.cpp:
```diff
@@ -4509,8 +4570,59 @@ class YarrGenerator final : public YarrJITInfo {
+        jit.move(areCanonicallyEquivalentCanonicalModeArgReg, GPRInfo::argumentGPR2);
+#endif
+
+        MacroAssembler::Call functionCall = jit.call(OperationPtrTag);
```
Let's do,

```
setupArguments<decltype(operationAreCanonicallyEquivalent)>(areCanonicallyEquivalentCharArgReg, areCanonicallyEquivalentPattCharArgReg, areCanonicallyEquivalentCanonicalModeArgReg);
callOperation<OperationPtrTag>(operationAreCanonicallyEquivalent);
```

msaboff replied:
Fixed.

Constellation reviewed (approved):
r=me

msaboff committed:
[JSC] Add JIT support for RegExp ignore case Backreferences in 16-bit patterns
https://bugs.webkit.org/show_bug.cgi?id=271617
rdar://125333123

Reviewed by Yusuke Suzuki.

Added code to call out to a new operationAreCanonicallyEquivalent() C++ helper, which wraps the inlined areCanonicallyEquivalent().
Given that the Yarr JIT has fixed assigned registers, added a thunk that saves all caller saves, moves the arguments from the
Yarr fixed registers to what operationAreCanonicallyEquivalent() expects and then calls it.  Upon return from the operation, we
restore registers, put the comparison return in a known register and return back to the Yarr JIT code that called out.

This change supports ARM64 and X86_64 (non-Windows) only.  Work should only be needed for the coding of the thunk generator for
other platforms with sufficient registers for general backrefernce processing in the Yarr JIT.

* Source/JavaScriptCore/yarr/YarrJIT.cpp:
(JSC::Yarr::JSC_DEFINE_JIT_OPERATION):
(JSC::Yarr::jitCompileInlinedTest):
* Source/JavaScriptCore/yarr/YarrJIT.h:
* Source/WTF/wtf/PlatformEnable.h:

Canonical link: https://commits.webkit.org/276681@main
